src/main/java/com/conveyal/datatools/manager/jobs/MergeLineContext.java

```diff
@@ -0,0 +1,899 @@
+                skipRecord = true;
+                return true;
+            }
+            // If a service_id from the active calendar has only the
+            // end_date in the future, the end_date shall be set to one
+            // day prior to the earliest start_date in future dataset
+            // before appending the calendar record to the merged file.
+            int endDateIndex = getFieldIndex(fieldsFoundInZip, "end_date");
+            if (index == endDateIndex) {
+                LocalDate endDate = LocalDate
+                    .parse(csvReader.get(endDateIndex), GTFS_DATE_FORMATTER);
+                if (!endDate.isBefore(futureFeedFirstDate)) {
+                    val = valueToWrite = futureFeedFirstDate
+                        .minus(1, ChronoUnit.DAYS)
+                        .format(GTFS_DATE_FORMATTER);
+                }
+            }
+        }
+        // Track service ID because we want to avoid removing trips that may reference this
+        // service_id when the service_id is used by calendar_dates that operate in the valid
+        // date range, i.e., before the future feed's first date.
+        if (field.name.equals(SERVICE_ID)) mergeFeedsResult.serviceIds.add(valueToWrite);
+        return false;
+    }
+
+    private boolean shouldUpdateFutureFeedStartDate(int startDateIndex) {
+        return index == startDateIndex &&
+            (
+                EXTEND_FUTURE == mergeFeedsResult.mergeStrategy ||
+                (
+                    CHECK_STOP_TIMES == mergeFeedsResult.mergeStrategy &&
+                        job.serviceIdsToExtend.contains(keyValue)
+                )
+            );
+    }
+
+    private void checkRoutesAndStopsIds(Set<NewGTFSError> idErrors) throws IOException {
+        // First, check uniqueness of primary key value (i.e., stop or route ID)
+        // in case the stop_code or route_short_name are being used. This
+        // must occur unconditionally because each record must be tracked
+        // by the reference tracker.
+        String primaryKeyValue = csvReader.get(table.getKeyFieldIndex(fieldsFoundInZip));
+        Set<NewGTFSError> primaryKeyErrors = referenceTracker
+            .checkReferencesAndUniqueness(primaryKeyValue, lineNumber, field, val, table);
+        // Merging will be based on route_short_name/stop_code in the active and future datasets. All
+        // matching route_short_names/stop_codes between the datasets shall be considered same route/stop. Any
+        // route_short_name/stop_code in active data not present in the future will be appended to the
+        // future routes/stops file.
+        if (useAltKey()) {
+            if (hasPrimaryKeyErrors(primaryKeyErrors)) {
+                // If alt key is empty (which is permitted), skip
+                // checking of alt key dupe errors/re-mapping values and
+                // simply use the primary key (route_id/stop_id).
+                skipRecord = true;
```

Comment:
I noticed, while revisiting this code, that this skipRecord should only occur if it has duplicate errors. The other checks (`"".equals(keyValue) && field.name.equals(table.getKeyFieldName()`) contained within hasPrimaryKeyErrors should be extracted and ultimately allow the record to be written to the output GTFS.

Reply:
Handled in https://github.com/ibi-group/datatools-server/pull/376/commits/e90b8c1e126267f0984a139725010523284e9136.